src/compiler/builder.ts

```diff
@@ -1132,26 +1135,37 @@ namespace ts {
             }
 
             // Emit only affected files if using builder for emit
-            if (!targetSourceFile && kind === BuilderProgramKind.EmitAndSemanticDiagnosticsBuilderProgram) {
-                // Emit and report any errors we ran into.
-                let sourceMaps: SourceMapEmitResult[] = [];
-                let emitSkipped = false;
-                let diagnostics: Diagnostic[] | undefined;
-                let emittedFiles: string[] = [];
-
-                let affectedEmitResult: AffectedFileResult<EmitResult>;
-                while (affectedEmitResult = emitNextAffectedFile(writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers)) {
-                    emitSkipped = emitSkipped || affectedEmitResult.result.emitSkipped;
-                    diagnostics = addRange(diagnostics, affectedEmitResult.result.diagnostics);
-                    emittedFiles = addRange(emittedFiles, affectedEmitResult.result.emittedFiles);
-                    sourceMaps = addRange(sourceMaps, affectedEmitResult.result.sourceMaps);
+            if (!targetSourceFile) {
+                if (kind === BuilderProgramKind.EmitAndSemanticDiagnosticsBuilderProgram) {
+                    // Emit and report any errors we ran into.
+                    let sourceMaps: SourceMapEmitResult[] = [];
+                    let emitSkipped = false;
+                    let diagnostics: Diagnostic[] | undefined;
+                    let emittedFiles: string[] = [];
+
+                    let affectedEmitResult: AffectedFileResult<EmitResult>;
+                    while (affectedEmitResult = emitNextAffectedFile(writeFile, cancellationToken, emitOnlyDtsFiles, customTransformers)) {
+                        emitSkipped = emitSkipped || affectedEmitResult.result.emitSkipped;
+                        diagnostics = addRange(diagnostics, affectedEmitResult.result.diagnostics);
+                        emittedFiles = addRange(emittedFiles, affectedEmitResult.result.emittedFiles);
+                        sourceMaps = addRange(sourceMaps, affectedEmitResult.result.sourceMaps);
+                    }
+                    return {
+                        emitSkipped,
+                        diagnostics: diagnostics || emptyArray,
+                        emittedFiles,
+                        sourceMaps
+                    };
+                }
+                else if (state.affectedFilesPendingEmitKind) {
+                    Debug.assert(kind === BuilderProgramKind.SemanticDiagnosticsBuilderProgram);
+                    // Remove files from affected files pending emit
+                    if (!emitOnlyDtsFiles || !find(state.affectedFilesPendingEmit!, (path, index) =>

```

Comment:
I think it's just me, but I'm having a really hard time reading this condition.  So we clear the list if we're not doing a dts-only emit (What's the alternative?  Since we're not in the emit-and-diagnostics case, is it no-emit?) or if the none of the unprocessed pending files did a non-dts emit?  Why are those conditions connected?

Reply:
We are cleaning affected files from pending emit if : 
- emitOnlyDeclaration is false
or
- affectedFilesPendingEmit has all files pending only in d.ts emit. I will update PR with more readable version